addons/point_of_sale/models/stock_picking.py

```diff
@@ -83,81 +84,114 @@ def _prepare_stock_move_vals(self, first_line, order_lines):
             'company_id': self.company_id.id,
         }
 
+    @api.model
+    def _create_move_lines_for_pos_order(self, moves, set_quantity_done_on_move=False):
+        moves._action_assign()
+        for move_line in moves.move_line_ids:
+            move_line.qty_done = move_line.product_uom_qty
+        mls_vals = []
+        moves_to_set = set()
+        for move in moves:
+            if float_compare(move.product_uom_qty, move.quantity_done, precision_rounding=move.product_uom.rounding) > 0:
+                remaining_qty = move.product_uom_qty - move.quantity_done
+                mls_vals.append(dict(move._prepare_move_line_vals(), qty_done=remaining_qty))
+                moves_to_set.add(move.id)
+        self.env['stock.move.line'].create(mls_vals)
+        if set_quantity_done_on_move:
+            for move in self.env['stock.move'].browse(moves_to_set):
+                move.quantity_done = move.product_uom_qty
+
+    def _create_production_lots_for_pos_order(self, lines):
+        ''' Search for existing lots and create missing ones.
+
+            :param lines: pos order lines with pack lot ids.
+            :type lines: pos.order.line recordset.
+
+            :return stock.production.lot recordset.
+        '''
+        self.ensure_one()
+        valid_lots = self.env['stock.production.lot']
+        if self.picking_type_id.use_existing_lots:
+            lots = lines.pack_lot_ids.filtered(lambda l: l.lot_name)
+            lots_data = set(lots.mapped(lambda l: (l.product_id.id, l.lot_name)))
+            existing_lots = self.env['stock.production.lot'].search([
+                ('company_id', '=', self.company_id.id),
+                ('product_id', 'in', lines.mapped('product_id').ids),
+                ('name', 'in', lots.mapped('lot_name')),
+            ])
+            # The previous search may return (product_id.id, lot_name) combinations that have no matching in lines.pack_lot_ids.
+            valid_lot_ids = set()
```

Comment:
here you assign a record set of lot instead of a set: valid_lot_ids = self.env['stock.production.lot'].

Reply:
@pimodoo Yes this is intented. Doing the union between two recordsets implies checking for unicity of their ids.
In some cases it might become a performance bottleneck when doing too many times (i.e. inside for loops on big
recordsets). Although I must admit that in this particular case the perf gain won't probably be much so if you
find working directly with recordsets easier to read, I'll change it.

Reply:
Thanks for the explanation, but I think that the code is more readable by working directly with recordset than having to create a set before. So as there is no big benefit to use sets, I'd prefer to work directly on recordsets